### PR TITLE
Generate string constants for enums

### LINKED ISSUE #30: Provide public const string values for enums.
Public const string values for enums would enable the use of switch expressions when mapping or doing other handling.

Current mapping code is one of:

```C#
if (source == ApiTaskStatus.Unclaimed)
	return TaskStatus.Unclaimed;
if (source == ApiTaskStatus.Pending)
	return TaskStatus.Pending;
if (source == ApiTaskStatus.Completed)
	return TaskStatus.Completed;
if (source == ApiTaskStatus.Rejected)
	return TaskStatus.Rejected;

throw new InvalidOperationException($"Unhandled task status: {source}");
```

or:

```C#
source switch
{
	_ when source == ApiTaskStatus.Unclaimed => TaskStatus.Unclaimed,
	_ when source == ApiTaskStatus.Pending => TaskStatus.Pending,
	_ when source == ApiTaskStatus.Completed => TaskStatus.Completed,
	_ when source == ApiTaskStatus.Rejected => TaskStatus.Rejected,
	_ => throw new InvalidOperationException($"Unhandled task status: {source}"),
}
```

Desired mapping code:

```C#
source switch
{
	ApiTaskStatus.Unclaimed => TaskStatus.Unclaimed,
	ApiTaskStatus.Pending => TaskStatus.Pending,
	ApiTaskStatus.Completed => TaskStatus.Completed,
	ApiTaskStatus.Rejected => TaskStatus.Rejected,
	_ => throw new InvalidOperationException($"Unhandled task status: {source}"),
}
```

### PR DESCRIPTION
**Pull request by jzbrooks**
Resolves #30

### CODE REVIEW

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 1]
```diff
@@ -252,6 +252,21 @@ private CodeGenFile GenerateEnum(ServiceEnumInfo enumInfo, Context context)
 
 						code.WriteLine();
 						code.WriteLine("readonly string m_value;");
+
+						code.WriteLine();
+						code.WriteLine("public static class Strings");
```
Comment (jzbrooks) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs:257]: I'm not sure how if an [obsolete] on the nest parent of the class is sufficient. Maybe this class needs one too?

Comment (jzbrooks) [reply src/Facility.CodeGen.CSharp/CSharpGenerator.cs:257]: Though, I suppose marking individual constants obsolete makes sense, because the constant should be obsolete if its enum variant is and a single (or subset) of the variants may be marked as such.

**Review by ejball — COMMENTED** [review]
I find it a bit unsettling to encourage a case-sensitive pattern when equality is otherwise case-insensitive.

If we care about that, the best solution I can think of is to normalize the case of the string in the constructor or in `ToString`. There are pros and cons for either one, and I'm not sure which is best.

To do that, we'd need a static read-only case-insensitive string cache (`HashSet<string>` in .NET Standard 2.1 or `Dictionary<string, string>` in .NET Standard 2.0) that used `StringComparer.OrdinalIgnoreCase` to look up the correct string. As an added bonus, it could be used to implement `IsDefined`.

**Comment by jzbrooks** [comment]
I believe the constants being distinct from the enum semantics this is actually a feature for two reasons:

1. It doesn't confuse the API by providing two mechanisms to do the same thing. Equality should typically be done by the comparison operator on the enum, not with the constants. This makes the constants a somewhat 'low level' API. If you want enum-ish semantics, use the enum constructor.
2. There's probably extra value in the simple thing. I'm reminded of https://jakewharton.com/the-economics-of-generated-code

It seems like a string cache might be something easily fixed in a later update if the need manifests.

It also occurred to me that initializing the enum properties with the static fields might make sense for DRY-ness. That probably matters less in a code generation context, but it _would_ strongly couple the standard enum variant properties to the string constant values.

All that said, I'm happy to do what you think best.

**Comment by ejball** [comment]
> Equality should typically be done by the comparison operator on the enum, not with the constants.

If the constant strings aren't for equality checking, I don't understand what they are for. Everything in the public API should have a clear purpose.

> probably extra value in the simple thing

If we can't explain what the extra value is, not adding it is even simpler.

> initializing the enum properties with the static fields

Yes, we should definitely do that; I should have noticed that in my review, thanks.

**Comment by jzbrooks** [comment]
I agree. I'll give the case-insensitivity strategy a whirl.

```csharp
Answer.Strings.Yes == answer.ToString()
Answer.Yes == answer
```
The semantic equivalence of these expressions seems awkward to me. Though, for and while loops are isomorphic, so duplicate syntactic mechanisms that express the same semantics aren't exactly uncommon…

---

A slight modification to the desired mapping code from the linked issue:

```csharp
string.ToLowerInvariant(source) switch
{
	ApiTaskStatus.Strings.Unclaimed => TaskStatus.Unclaimed,
	ApiTaskStatus.Strings.Pending => TaskStatus.Pending,
	ApiTaskStatus.Strings.Completed => TaskStatus.Completed,
	ApiTaskStatus.Strings.Rejected => TaskStatus.Rejected,
	_ => throw new InvalidOperationException($"Unhandled task status: {source}"),
}
```

In the case of the service I work on, we _might_ be able to get away without `string.ToLowerInvariant` call because all of the clients applications we care about use facility-based clients (so all of the constants there would naturally have the appropriate lowercase form), so we could selectively normalize during mapping. Granted, that might be playing with fire for no good reason. My thought is that it doesn't put a huge burden on consuming code. At worst its just a little arcane knowledge, but it doesn't seem all that arcane to me.

Just to put that out there. It's probably an esoteric use-case.

---

> If we can't explain what the extra value is, not adding it is even simpler.

That's fair enough. 😅

**Comment by jzbrooks** [comment]
I'm coming to realize I don't understand the value of a cache here. Why does it not make sense to just lowercase the value passed into the constructor and be done with it? Avoiding extra allocation?

If the cache is going to be updated when unseen values are encountered, we couldn't use it for `IsDefined`, right? Because the cache may contain undefined values.

Are you thinking the cache would be a read only dictionary initialized with defined values? We could just call `ToLowerInvariant` on undefined values to m_value consistently cased.

File: src/Facility.ConformanceApi/Answer.g.cs [hunk 2]
```diff
@@ -17,25 +17,43 @@ namespace Facility.ConformanceApi
 	[JsonConverter(typeof(AnswerJsonConverter))]
 	public partial struct Answer : IEquatable<Answer>
 	{
+		private static readonly Dictionary<string, string> s_normalizedConstants = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```
Comment (ejball) [inline src/Facility.ConformanceApi/Answer.g.cs:20]: This should be moved below `s_values` and be initialized similarly (not with a static constructor).

File: src/Facility.ConformanceApi/Answer.g.cs [hunk 3]
```diff
@@ -17,25 +17,43 @@ namespace Facility.ConformanceApi
...
 		/// <summary>
 		/// Creates an instance.
 		/// </summary>
-		public Answer(string value) => m_value = value;
+		public Answer(string value)
+		{
+			if (!s_normalizedConstants.TryGetValue(value, out var normalizedValue))
+			{
+				normalizedValue = value.ToLowerInvariant();
+				s_normalizedConstants[normalizedValue] = normalizedValue;
```
Comment (ejball) [inline src/Facility.ConformanceApi/Answer.g.cs:52]: This would not be thread-safe, but we don't need to do it anyway.

File: src/Facility.ConformanceApi/Answer.g.cs [hunk 4]
```diff
@@ -45,7 +63,7 @@ public partial struct Answer : IEquatable<Answer>
 		/// <summary>
 		/// Checks for equality.
 		/// </summary>
-		public bool Equals(Answer other) => StringComparer.OrdinalIgnoreCase.Equals(ToString(), other.ToString());
+		public bool Equals(Answer other) => StringComparer.Ordinal.Equals(ToString(), other.ToString());
```
Comment (ejball) [inline src/Facility.ConformanceApi/Answer.g.cs:66]: Non-normalized values should still be compared case-insensitively, so these should be reverted.

File: src/Facility.ConformanceApi/Answer.g.cs [hunk 5]
```diff
@@ -17,25 +17,43 @@ namespace Facility.ConformanceApi
...
 
 		/// <summary>
 		/// Creates an instance.
 		/// </summary>
-		public Answer(string value) => m_value = value;
+		public Answer(string value)
+		{
+			if (!s_normalizedConstants.TryGetValue(value, out var normalizedValue))
+			{
+				normalizedValue = value.ToLowerInvariant();
```
Comment (ejball) [inline src/Facility.ConformanceApi/Answer.g.cs:51]: We should just set `m_value` to the normalized value, not also convert it to lowercase.

**Review by ejball — COMMENTED** [review]
> Why does it not make sense to just lowercase the value passed into the constructor and be done with it? Avoiding extra allocation?

And because we should preserve the desired case as indicated by the FSD.

> If the cache is going to be updated when unseen values are encountered

We should not update the cache with undefined values.

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs [hunk 6]
```diff
@@ -278,17 +292,6 @@ private CodeGenFile GenerateEnum(ServiceEnumInfo enumInfo, Context context)
 							}
 						}
 
-						code.WriteLine();
-						code.WriteLine($"private static readonly Dictionary<string, string> s_normalizedConstants = new Dictionary<string, string>");
```
Comment (ejball) [inline src/Facility.CodeGen.CSharp/CSharpGenerator.cs]: This seems right, except it needs `(StringComparer.OrdinalIgnoreCase)`. Why this change?

**Comment by jzbrooks** [comment]
I thought better about that, but the airplane WiFi cut out before I could push. I'll get it sorted shortly. 

I ran into some initialization order quirks that I worked around by adding a couple of static methods on the Strings class. There may be a better way.

File: src/Facility.ConformanceApi/Answer.g.cs (hunk 3 above)
Comment (jzbrooks) [reply src/Facility.ConformanceApi/Answer.g.cs:52]: That was the main impetus behind me thinking better of it. 😬

File: src/Facility.CodeGen.CSharp/CSharpGenerator.cs (hunk 6 above)
Comment (jzbrooks) [reply src/Facility.CodeGen.CSharp/CSharpGenerator.cs]: It was a half-baked idea, but the rationale was `m_value` would always be normalized (which I get doesn't make sense now).

**Comment by jzbrooks** [comment]
It turns out that C# initialization order forces our hand in using the `ToString` approach. The static dictionary field isn't initialized in the constructor.

File: src/Facility.ConformanceApi/Answer.g.cs (hunk 2 above)
Comment (jzbrooks) [reply src/Facility.ConformanceApi/Answer.g.cs]: FWIW, this was a remnant of an attempt to deduce C#'s initialization order semantics without documentation. It's gone now.

**Review by ejball — CHANGES_REQUESTED** [review]
> C# initialization order forces our hand

Ah, yes, the problem is that we define static properties that call the constructor... Well, the `ToString` approach is probably better anyway. :-)

**Review by ejball — APPROVED** [review]
Thanks!
